File: src/lib/ipc/picus/tx.ts
/**
 * Picus explicit transactions — `BEGIN`, then a decision that is the user's.
 *
 * Four calls, and the shape of them is deliberate: three of them **answer with the
 * state the connection is in afterwards**, and the fourth asks for that state on its
 * own. Nothing on this side keeps a flag. The engine is the only thing that knows
 * whether a transaction is open, because it is the only thing that can end one by
 * itself — a statement that fails inside a block leaves PostgreSQL in an *aborted*
 * transaction which refuses everything until it is rolled back, and a client-side
 * boolean would happily go on saying "active" while the connection accepted nothing.
 *
 * Which is why {@link txState} is asked after every statement rather than only after
 * a transaction verb: an open transaction changes what the next statement means, and
 * a failed one changes whether there will be a next statement at all.
 */

import { picus } from '../rpc';

/**
 * Where a connection's transaction stands.
 *
 * `savepoint` is part of the contract but PostgreSQL never reports it: the server's
 * transaction status does not distinguish a block with savepoints from one without,
 * so the provider does not guess. An engine that tracks its own savepoints can.
 */
export type TxState = 'none' | 'active' | 'failed' | 'savepoint';

/** The answer to begin / commit / rollback. */
export interface TxOutcome {
  /** The state the connection is in **after** the call — read back from the server. */
  state: TxState;
  /** What happened, in the user's terms. Shown verbatim. */
  message: string;
  /**
   * Statements that ran inside the transaction that just ended.
   *
   * `0` means **not counted**, never "none ran": the PostgreSQL provider's
   * transaction calls see a connection and no statements, so only a session that
   * counts them can fill this in. Anything rendering it must treat zero as absent.
   */
  statements: number;
}

/** Open an explicit transaction. Rejects, in words, when one is already open. */
export function txBegin(id: string): Promise<TxOutcome> {
  return picus('picus_tx_begin', { id });
}

/**
 * Commit the open transaction.
 *
 * Rejects on a **failed** transaction rather than forwarding it: PostgreSQL's own
 * `COMMIT` inside an aborted block performs a rollback and reports success, which
 * would tell the user their work was saved at the moment it was thrown away.
 */
export function txCommit(id: string): Promise<TxOutcome> {
  return picus('picus_tx_commit', { id });
}

/**
 * Roll the open transaction back — including, and especially, a failed one.
 *
 * The only call here that works in every state, which is what makes it safe to put
 * behind "close anyway". Rolling back with nothing open succeeds and says so.
 */
export function txRollback(id: string): Promise<TxOutcome> {
  return picus('picus_tx_rollback', { id });
}

/** Ask the server where the transaction stands. Cheap: one round trip, no writes. */
export function txState(id: string): Promise<TxState> {
  return picus('picus_tx_state', { id });
}

File: crates/products/picus/be/src/tx.rs
//! `tx` domain — explicit transactions on an open connection.
//!
//! Four verbs over the seam, and the shape of them is the point: `begin`, `commit`
//! and `rollback` each answer with the state the connection is **in afterwards**,
//! read back from the server, and [`picus_tx_state`] asks the same question on its
//! own. Nothing in this process remembers whether a transaction is open — the engine
//! is the only thing that knows, because it is the only thing that can abort one on
//! its own (a statement that fails inside a block puts PostgreSQL into an aborted
//! transaction that refuses everything until it is ended).
//!
//! The interface calls [`picus_tx_state`] after every statement for exactly that
//! reason: an open transaction changes what the *next* statement means, and a failed
//! one changes whether there will be a next statement at all. Both facts have to be
//! on screen before the user writes it, not after.
//!
//! An engine without transactions answers
//! [`DbError::unsupported`](picus_db_api::prelude::DbError::unsupported) from the
//! trait's own default — but the interface should never get that far: it reads
//! `capabilities.transactions.supported` from the descriptor and does not offer what
//! the engine does not have.

use picus_core::prelude::PicusState;
use picus_db_api::prelude::{TxOutcome, TxState};

use crate::connections::require_session;

/// Open an explicit transaction. Fails, in words, when one is already open.
#[arbor_rpc::handler]
async fn picus_tx_begin(state: &PicusState, id: String) -> Result<TxOutcome, String> {
    require_session(state, &id)?.begin().await.map_err(|e| e.to_string())
}

/// Commit the open transaction.
///
/// Refused on a failed one rather than forwarded, and that refusal is the engine's
/// to explain: on PostgreSQL a `COMMIT` inside an aborted block *performs a
/// rollback* and reports success, so passing it through would tell the user their
/// work was saved at the moment it was discarded.
#[arbor_rpc::handler]
async fn picus_tx_commit(state: &PicusState, id: String) -> Result<TxOutcome, String> {
    require_session(state, &id)?.commit().await.map_err(|e| e.to_string())
}

/// Roll the open transaction back.
///
/// The one call here that must work in every state the others refuse — a failed
/// transaction is precisely the situation where it is the only thing left. Rolling
/// back when nothing is open is success with a message saying so, not an error:
/// this is also the call the interface makes on the way out of a window, and a
/// close that fails because there was nothing to undo would be a close nobody
/// completes.
#[arbor_rpc::handler]
async fn picus_tx_rollback(state: &PicusState, id: String) -> Result<TxOutcome, String> {
    require_session(state, &id)?.rollback().await.map_err(|e| e.to_string())
}

/// Where the transaction stands, asked of the server.
#[arbor_rpc::handler]
async fn picus_tx_state(state: &PicusState, id: String) -> Result<TxState, String> {
    require_session(state, &id)?.tx_state().await.map_err(|e| e.to_string())
}
